projects/evm-compatible-bridge/federator.md

```diff
@@ -0,0 +1,300 @@
+
+The federator service will keep track of events on both sides of the bridge and vote on the actions to be taken.
+This will be done by watching events on the federation contract on the EVM chain and polling events from the coordinator service on Hathor.
+
+Since Hathor does not currently have a smart contract implementation, we will use a MultiSig address to check that a certain number of federators have consented to any transaction.
+To avoid separate instances of the MultiSig wallet to propose conflicting transactions, we will use a coordinator service to choose the utxos and collect signatures from the federators.
+
+Each federator will have 2 loops, one to check events on the EVM compatible chain (i.e. EVM loop) and another to check events on the coordinator service (i.e. polling loop).
+The EVM loop will get the events from the chain, this loop will have events of transaction requests from the EVM compatible chain to Hathor and confirmations on transactions.
+The polling loop will check the coordinator service polling api and vote to perform the actions as requested.
+
+## Federator service
+
+### EVM loop
+
+We need to get events from the bridge contract, wait a pre-configured number of blocks (to increase the confirmation level of the transaction) and send the transaction on Hathor as requested.
+There should be 3 types of events:
+
+- A request to cross the token from the EVM chain to Hathor
+  - If the token is native to the EVM chain, an equivalent Hathor token should already exist and the MultiSig wallet should own the authority utxos to it.
+  - If the token is native to the Hathor chain, the equivalent token is already burned and we need to unlock the tokens on Hathor.
+  - The loop will have to use the coordinator service to send the amount of tokens requested from the MultiSig wallet to the destination address.
+- A request to support a new token native to the EVM chain.
+  - We need to check if this is a duplicate request
+  - We need to use the coordinator service to create of the equivalent token on Hathor.
+  - We need to use the bridge smart contract to save the Hathor equivalent token uid once the token is created.
+- Confirmation that an operation is complete
+  - This way we can clean any state pertaining to this crossing of tokens and mark it as complete.
+  - If the crossing was originated in Hathor, we need to send a request to the coordinator service so it knows the crossing is over.
+
+### Polling loop
+
+We will pool for events, which are requests for voting, and vote on them.
+If the event requires a vote on the smart contract, we will send a transaction to the EVM chain with the vote.
+If the event requires a vote on the MultiSig wallet, we will send a message to the coordinator service with a signature.
+
+The federator will be responsible to check the data on the event, it will have its own headless wallet to check the information from the Hathor network and it can read the data from the EVM chain.
```

Comment:
Yes, this is very important, all transactions and events should be confirmed by reading the on-chain data.
This is to prevent a communication error or a malicious party to cause any damages to the service.

Reply:
Can you detail which checks need to be done? Other than the regular information, how will the federator know this crossing hasn't been done before? I mean, how to prevent duplicated requests?